commons/ava_testsuite/staking_gecko_network_tests.go

```diff
@@ -191,21 +179,266 @@ func (test FiveNodeStakingNetworkGetValidatorsTest) GetTimeout() time.Duration {
+		context.Fatal(stacktrace.Propagate(err, "Could not remove the first service with duped node ID"))
+	}
+	delete(allServiceIds, badServiceId1)
+	delete(allGeckoClients, badServiceId1)
+	delete(allNodeIds, badServiceId1)
+	logrus.Info("Successfully removed first node with duplicate ID, leaving only the second")
+
+	// Now that the first duped node is gone, verify that the original node is still connected to just boot nodes and
+	//  the second duped-ID node is now accepted by the boot nodes
+	logrus.Info("Verifying that the network has connected to the second node with a previously-duplicated node ID...")
+	for serviceId, _ := range allServiceIds {
+		acceptableNodeIds := make(map[string]bool)
+
+		// All nodes should have the boot nodes (though a boot node won't have itself)
+		for bootServiceId, _ := range bootServiceIds {
+			if serviceId != bootServiceId {
+				bootNodeId := allNodeIds[bootServiceId]
+				acceptableNodeIds[bootNodeId] = true
+			}
+		}
+
+		// Boot nodes should have all nodes
+		if _, found := bootServiceIds[serviceId]; found {
+			acceptableNodeIds[allNodeIds[NODE_SERVICE_ID]] = true
+			acceptableNodeIds[badServiceNodeId2] = true
+			verifyExpectedPeers(context, serviceId, allGeckoClients[serviceId], acceptableNodeIds, len(allServiceIds) - 1, false)
+		} else {
+			verifyExpectedPeers(context, serviceId, allGeckoClients[serviceId], acceptableNodeIds, len(bootServiceIds), false)
+		}
+	}
+	logrus.Info("Verified that the network has settled on the second node with previously-duplicated ID")
+}
+
+func (f FiveNodeStakingNetworkDuplicateIdTest) GetNetworkLoader() (testsuite.TestNetworkLoader, error) {
+	return getFiveNodeStakingLoader()
+}
+
+func (f FiveNodeStakingNetworkDuplicateIdTest) GetTimeout() time.Duration {
+	return 120 * time.Second
+}
+
 // =============== Helper functions =============================
 
 // TODO TODO TODO Rename this
 func getFiveNodeStakingLoader() (testsuite.TestNetworkLoader, error) {
 	serviceConfigs := map[int]ava_networks.TestGeckoNetworkServiceConfig{
-		NODE_CONFIG_ID: *ava_networks.NewTestGeckoNetworkServiceConfig(true, ava_services.LOG_LEVEL_DEBUG),
+		NORMAL_NODE_CONFIG_ID: *ava_networks.NewTestGeckoNetworkServiceConfig(true, ava_services.LOG_LEVEL_DEBUG),
+		SAME_CERT_CONFIG_ID:   *ava_networks.NewTestGeckoNetworkServiceConfig(false, ava_services.LOG_LEVEL_DEBUG),
 	}
 	return ava_networks.NewTestGeckoNetworkLoader(
 		ava_services.LOG_LEVEL_DEBUG,
 		true,
 		serviceConfigs,
 		map[int]int{
-			NODE_SERVICE_ID: NODE_CONFIG_ID,
+			NODE_SERVICE_ID: NORMAL_NODE_CONFIG_ID,
 		},
 		2,
 		2)
 }
 
+/*
+Verifies that a node's actual peers are what we expect
+
+Args:
+	context: Test context (used for failing if there's a problem)
+	serviceId: Service ID of the node whose peers are being examined
+	client: Gecko client for the node being examined
+	acceptableNodeIds: A "set" of acceptable node IDs where, if a peer doesn't have this ID, the test will be failed
```

Comment:
I think this one doesn't play well with the bootstrappers because they represent themselves in their own peerlist by a random ID. the condition we put on them is that "everyone else is in my peer list, and i might have a peer that is a wonky id (that peer is me)"

Reply:
although for now its fine because we're not using this function in fully connected test, but if you were to, i think we'd have to examine that

Reply:
I actually skip the bootstrapper itself. But even if we added it, it'd be okay because it's just the list of what's acceptable - it's not a strict requirement that all peers be in that list. And I have to have it that way, because for my dupe node ID test it's unknown what their peer list will actually look like - we can only assert what subset the peer elements should be a part of

Reply:
ah so bootstrappers are never analyzed in this function?

" it's not a strict requirement that all peers be in that list" i think the code does require that, if i'm reading it right. it iterates through every peer in the peer list, and fails if one of those peers is not in the accepted node list

Reply:
Hey @galenmarchetti are you sure that bootstrappers have their own ID in the peers list? the behaviour I've been seeing, and which is how I wrote my test, is that a bootstrapper won't have its own ID in its peer list. This aligns with what I'd seen when I was pinging Docker containers, and the test seems to pass with this assumption